PDU Fix

mikepj opened the pull request:
While I was able to see the PDU objects being collected on the `inputunifi` side, I wasn't seeing them in the `influxunifi` output.  I traced the problem down to the PDU metrics not being added to the devices array in the `inputunifi` collector.  I am now seeing PDU statistics showing up in my Grafana/Influxdb setup.

I'm thinking since the fix was on the input side, the other output packages should also be able to pull the same data now without further modifications, but I didn't verify that was the case.

KetchupBomb commented:
I pulled your repo and checked out your branch, @mikepj, and it panics:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0xe4c2d7]

goroutine 13 [running]:
github.com/prometheus/client_golang/prometheus.NewConstMetric(0x0?, 0xc0002e4ea8?, 0x7584c5?, {0xc000354960?, 0xf0eca0?, 0xc000314701?})
        /go/pkg/mod/github.com/prometheus/client_golang@v1.16.0/prometheus/value.go:105 +0x17
github.com/prometheus/client_golang/prometheus.MustNewConstMetric(...)
        /go/pkg/mod/github.com/prometheus/client_golang@v1.16.0/prometheus/value.go:126
github.com/unpoller/unpoller/pkg/promunifi.(*Report).export(0xf61160?, 0xc0000a6660?, 0x0?)
        /host/pkg/promunifi/report.go:65 +0x3c
github.com/unpoller/unpoller/pkg/promunifi.(*promUnifi).exportMetrics(0xc0003540c0, {0x11cf6c0?, 0xc000298280}, 0x0?, 0x0?)
        /host/pkg/promunifi/collector.go:339 +0x20d
created by github.com/unpoller/unpoller/pkg/promunifi.(*promUnifi).collect in goroutine 52
        /host/pkg/promunifi/collector.go:322 +0x31b
```

I am not positive on how it should be built. I simply `go build` inside a recent [Golang image](https://hub.docker.com/_/golang). It builds a binary with:

```
# md5sum unpoller 
515b0eb3fcecf330c854e4fd9c8e0d1d  unpoller
```

Building on `master` produces a binary with:

```
# md5sum unpoller 
5e3e08ede20ffc98a4760bf75e48b86c  unpoller
```

`master` does not panic (PDU metrics are still missing).

mikepj commented:
I had zero knowledge of Go before starting to look into this, but `go build` was all I needed to compile a build from this branch.  From your stacktrace, It looks like there is a problem in the prometheus output package.  I was testing with influxdb, so I wouldn't have run into it.  I don't have a prometheus installation here to test with, but maybe just enabling that output plugin with a bogus server config would be enough to run into that crash.  

It also looks like the unit tests are failing due to the metrics counts not matching up (more metrics are being generated than previously expected).  Maybe there is PDU content in the test data that is now making it through to the output and throwing off the numbers.  @platinummonkey do you know if there's PDU test data in the mix?

platinummonkey commented:
Yeah I can also take a look this week. The output tests are new to check for changes like what’s happened with int/float changes by ubiquiti. It does dump in the test logs what the yaml should be, to make it less onerous.

mikepj commented:
@KetchupBomb I setup a prometheus server and was able to replicate the crashing bug.  The problem ended up being missing metrics descriptions.  I updated the code that provides metric descriptions and it's working fine on my end now.  Give it a shot to see if that fixes the problem for you.

@platinummonkey I updated the test expectations, so the tests are passing on my end.

KetchupBomb commented:
@mikepj, with your subsequent commits I am now able to see PDU metrics when scraping!

```
# curl -s localhost:9130/metrics | grep ^unpoller | cut -d'{' -f1 | sort | uniq -c | sort -nk1
      1 unpoller_device_outlet_ac_power_consumption
     20 unpoller_device_outlet_outlet_caps
     20 unpoller_device_outlet_outlet_current
     20 unpoller_device_outlet_outlet_power
     20 unpoller_device_outlet_outlet_power_factor
     20 unpoller_device_outlet_outlet_voltage
```

platinummonkey commented:
💯 thanks for the fix!! Will cut a release

platinummonkey commented:
https://github.com/unpoller/unpoller/releases/tag/v2.9.0
